### ISSUE TITLE
[Lowest Price Node] No price data when using dynamic configuration

### ISSUE THREAD

**Issue by gulli1986** [ottopaulsen/node-red-contrib-power-saver#59]
Hello,

I am currently using a dynamic configuration on the `lowest price` node to adjust the amount of hours ("hours on") based on the state of charge of my car. It has been working fine so long however I discovered today that I sometimes get an error message after inputing the dynamic configuration to the node;

![nodered](https://user-images.githubusercontent.com/84970544/173201193-e4c0b60d-2f07-48f1-90cd-ac4443778115.PNG)

It seems like the `lowest price` node "forgets" the saved prices once the dynamic configuration is sent to the node.
I just updated to `Power Saver [v3.5.5]`, can this be related?

It looks like the error is showing after some time, sometimes it works, sometimes it doesn't.

**Comment by ottopaulsen** [comment]
So when you send one config change, you get first `Using saved prices` and then `No price data`, is that correct?

If you upgraded recently, please try to delete the node and create it again, just to make sure there is nothing wrong in the nodes context. Alternatively delete all the context data and keep the node. You find the context data by selecting the node, then select `Context Data` where you see the debug window, then click the refresh button for Node. You should see data for `context`, `lastPlan`, `lastPriceData`, `lastSource` and a coupe of dates. It is safe to delete everything. 

See if this helps first.
If not, copy config and paste it here. Also paste the message you use to change config.

**Comment by gulli1986** [comment]
Hello Otto @ottopaulsen. Thanks for the help. I tried both removing the node and deleting context data without any help. That is correct as you mention, I first get `Using saved prices` then immediately `No price data`.

Here is my configuration:
`[{"id":"96708db28a6dbdc3","type":"tab","label":"Flow 1","disabled":false,"info":"","env":[]},{"id":"031b05ea780117d5","type":"debug","z":"96708db28a6dbdc3","name":"Debug 03","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"true","targetType":"full","statusVal":"","statusType":"auto","x":960,"y":200,"wires":[]},{"id":"2ca238794cf3d9ec","type":"api-current-state","z":"96708db28a6dbdc3","name":"Battery Remaining Kia","server":"6fd34f.07042cb","version":3,"outputs":1,"halt_if":"","halt_if_type":"str","halt_if_compare":"is","entity_id":"input_number.remaining_battery_kia","state_type":"str","blockInputOverrides":false,"outputProperties":[{"property":"payload","propertyType":"msg","value":"","valueType":"entityState"},{"property":"topic","propertyType":"msg","value":"battery_remaining","valueType":"str"}],"for":"0","forType":"num","forUnits":"minutes","override_topic":false,"state_location":"payload","override_payload":"msg","entity_location":"data","override_data":"msg","x":440,"y":360,"wires":[["dd71c72be5f2cf58"]]},{"id":"65588d5ba85e5323","type":"api-current-state","z":"96708db28a6dbdc3","name":"Charging Time Start","server":"6fd34f.07042cb","version":3,"outputs":1,"halt_if":"","halt_if_type":"str","halt_if_compare":"is","entity_id":"input_datetime.charging_time_start","state_type":"str","blockInputOverrides":false,"outputProperties":[{"property":"payload","propertyType":"msg","value":"","valueType":"entityState"},{"property":"topic","propertyType":"msg","value":"start_time","valueType":"str"}],"for":"0","forType":"num","forUnits":"minutes","override_topic":false,"state_location":"payload","override_payload":"msg","entity_location":"data","override_data":"msg","x":440,"y":420,"wires":[["2d06c424c14bb626"]]},{"id":"b266ee42a44237bf","type":"api-current-state","z":"96708db28a6dbdc3","name":"Charging Time End","server":"6fd34f.07042cb","version":3,"outputs":1,"halt_if":"","halt_if_type":"str","halt_if_compare":"is","entity_id":"input_datetime.charging_time_end","state_type":"str","blockInputOverrides":false,"outputProperties":[{"property":"payload","propertyType":"msg","value":"","valueType":"entityState"},{"property":"topic","propertyType":"msg","value":"end_time","valueType":"str"}],"for":"0","forType":"num","forUnits":"minutes","override_topic":false,"state_location":"payload","override_payload":"msg","entity_location":"data","override_data":"msg","x":430,"y":480,"wires":[["2d06c424c14bb626"]]},{"id":"2d06c424c14bb626","type":"join","z":"96708db28a6dbdc3","name":"","mode":"custom","build":"object","property":"payload","propertyType":"msg","key":"topic","joiner":"\\n","joinerType":"str","accumulate":false,"timeout":"","count":"3","reduceRight":false,"reduceExp":"","reduceInit":"","reduceInitType":"","reduceFixup":"","x":910,"y":420,"wires":[["d7cca40f0874b8d2"]]},{"id":"d7cca40f0874b8d2","type":"function","z":"96708db28a6dbdc3","name":"Dyn. Config","func":"msg.payload = {\n    \"config\":{\n        \"fromTime\":msg.payload.start_time,\n        \"toTime\" : msg.payload.end_time,\n        \"hoursOn\" : msg.payload.hours_to_charge,\n        \"doNotSplit\" : false,\n        \"sendCurrentValueWhenRescheduling\" : true,\n        \"outputIfNoSchedule\" : false,\n        \"outputOutsidePeriod\": false\n       }\n}\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":1050,"y":420,"wires":[["7d5ff290b6710ceb"]]},{"id":"7d5ff290b6710ceb","type":"change","z":"96708db28a6dbdc3","name":"","rules":[{"t":"delete","p":"topic","pt":"msg"}],"action":"","property":"","from":"","to":"","reg":false,"x":1220,"y":420,"wires":[["7e0823a129a8ce03"]]},{"id":"dd71c72be5f2cf58","type":"function","z":"96708db28a6dbdc3","name":"Calculate Hours to Charge","func":"battery_int = parseInt(msg.payload);\nhours_to_charge_based_on_remaining_battery = Math.ceil ( (( 30 * (1 -  (battery_int / 100))) / 6.1 ) );\n\n\nmsg.payload = hours_to_charge_based_on_remaining_battery,\nmsg.topic = \"hours_to_charge\"\n\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":700,"y":360,"wires":[["2d06c424c14bb626"]]},{"id":"68ccca4e0c2ef850","type":"comment","z":"96708db28a6dbdc3","name":"Dynamic configuration for Kia charging","info":"","x":170,"y":320,"wires":[]},{"id":"07b99a28146b6326","type":"api-call-service","z":"96708db28a6dbdc3","name":"Stop Charging","server":"6fd34f.07042cb","version":5,"debugenabled":false,"domain":"easee","service":"pause","areaId":[],"deviceId":[],"entityId":[],"data":"{\"charger_id\":\"EHDH7X59\"}","dataType":"json","mergeContext":"","mustacheAltTags":false,"outputProperties":[],"queue":"none","x":1200,"y":180,"wires":[[]]},{"id":"47448edbd9c76c15","type":"api-call-service","z":"96708db28a6dbdc3","name":"Resume Charging","server":"6fd34f.07042cb","version":5,"debugenabled":false,"domain":"easee","service":"resume","areaId":[],"deviceId":[],"entityId":[],"data":"{\"charger_id\":\"EHDH7X59\"}","dataType":"json","mergeContext":"","mustacheAltTags":false,"outputProperties":[],"queue":"none","x":990,"y":80,"wires":[[]]},{"id":"8e3d23984ae7a815","type":"server-state-changed","z":"96708db28a6dbdc3","name":"Smart Charging Button Pressed","server":"6fd34f.07042cb","version":4,"exposeToHomeAssistant":false,"haConfig":[{"property":"name","value":""},{"property":"icon","value":""}],"entityidfilter":"input_button.update_smart_charging","entityidfiltertype":"exact","outputinitially":false,"state_type":"str","haltifstate":"","halt_if_type":"str","halt_if_compare":"is","outputs":1,"output_only_on_state_change":true,"for":"0","forType":"num","forUnits":"minutes","ignorePrevStateNull":false,"ignorePrevStateUnknown":false,"ignorePrevStateUnavailable":false,"ignoreCurrentStateUnknown":false,"ignoreCurrentStateUnavailable":false,"outputProperties":[{"property":"payload","propertyType":"msg","value":"","valueType":"entityState"},{"property":"data","propertyType":"msg","value":"","valueType":"eventData"},{"property":"topic","propertyType":"msg","value":"","valueType":"triggerId"}],"x":150,"y":420,"wires":[["2ca238794cf3d9ec","65588d5ba85e5323","b266ee42a44237bf"]]},{"id":"69332bdf051a776d","type":"server-state-changed","z":"96708db28a6dbdc3","name":"Charging Cable Connected","server":"6fd34f.07042cb","version":4,"exposeToHomeAssistant":false,"haConfig":[{"property":"name","value":""},{"property":"icon","value":""}],"entityidfilter":"sensor.driveway_status","entityidfiltertype":"exact","outputinitially":false,"state_type":"str","haltifstate":"disconnected","halt_if_type":"str","halt_if_compare":"is_not","outputs":2,"output_only_on_state_change":true,"for":"0","forType":"num","forUnits":"minutes","ignorePrevStateNull":false,"ignorePrevStateUnknown":false,"ignorePrevStateUnavailable":false,"ignoreCurrentStateUnknown":false,"ignoreCurrentStateUnavailable":false,"outputProperties":[{"property":"payload","propertyType":"msg","value":"","valueType":"entityState"},{"property":"data","propertyType":"msg","value":"","valueType":"eventData"},{"property":"topic","propertyType":"msg","value":"","valueType":"triggerId"}],"x":130,"y":120,"wires":[["aa747b0e13dc8092"],[]]},{"id":"aa747b0e13dc8092","type":"api-current-state","z":"96708db28a6dbdc3","name":"Nordpool Sensor","server":"6fd34f.07042cb","version":3,"outputs":1,"halt_if":"","halt_if_type":"str","halt_if_compare":"is","entity_id":"sensor.nordpool_kwh_oslo_nok_3_10_025","state_type":"str","blockInputOverrides":false,"outputProperties":[{"property":"payload","propertyType":"msg","value":"","valueType":"entityState"},{"property":"data","propertyType":"msg","value":"","valueType":"entity"}],"for":"0","forType":"num","forUnits":"minutes","override_topic":false,"state_location":"payload","override_payload":"msg","entity_location":"data","override_data":"msg","x":370,"y":140,"wires":[["9c4e0582aaf7bb79"]]},{"id":"9c4e0582aaf7bb79","type":"ps-receive-price","z":"96708db28a6dbdc3","name":"Price Receiver","x":560,"y":140,"wires":[["7e0823a129a8ce03"]]},{"id":"728873ed6cc85028","type":"comment","z":"96708db28a6dbdc3","name":"02 - Power Saver Kia Soul Charging","info":"","x":160,"y":60,"wires":[]},{"id":"ff025927d417b28f","type":"api-current-state","z":"96708db28a6dbdc3","name":"Charge Now?","server":"6fd34f.07042cb","version":3,"outputs":2,"halt_if":"on","halt_if_type":"str","halt_if_compare":"is","entity_id":"input_boolean.charge_now","state_type":"str","blockInputOverrides":false,"outputProperties":[{"property":"payload","propertyType":"msg","value":"","valueType":"entityState"},{"property":"data","propertyType":"msg","value":"","valueType":"entity"}],"for":"0","forType":"num","forUnits":"minutes","override_topic":false,"state_location":"payload","override_payload":"msg","entity_location":"data","override_data":"msg","x":980,"y":140,"wires":[["4f4375ddb94a7a8f"],["07b99a28146b6326"]]},{"id":"4f4375ddb94a7a8f","type":"api-call-service","z":"96708db28a6dbdc3","name":"Resume Charging","server":"6fd34f.07042cb","version":5,"debugenabled":false,"domain":"easee","service":"resume","areaId":[],"deviceId":[],"entityId":[],"data":"{\"charger_id\":\"EHDH7X59\"}","dataType":"json","mergeContext":"","mustacheAltTags":false,"outputProperties":[],"queue":"none","x":1210,"y":100,"wires":[[]]},{"id":"dad8798ccd566365","type":"server-state-changed","z":"96708db28a6dbdc3","name":"Charge Now","server":"6fd34f.07042cb","version":4,"exposeToHomeAssistant":false,"haConfig":[{"property":"name","value":""},{"property":"icon","value":""}],"entityidfilter":"input_boolean.charge_now","entityidfiltertype":"exact","outputinitially":false,"state_type":"str","haltifstate":"on","halt_if_type":"str","halt_if_compare":"is","outputs":2,"output_only_on_state_change":true,"for":"0","forType":"num","forUnits":"minutes","ignorePrevStateNull":false,"ignorePrevStateUnknown":false,"ignorePrevStateUnavailable":false,"ignoreCurrentStateUnknown":false,"ignoreCurrentStateUnavailable":false,"outputProperties":[{"property":"payload","propertyType":"msg","value":"","valueType":"entityState"},{"property":"data","propertyType":"msg","value":"","valueType":"eventData"},{"property":"topic","propertyType":"msg","value":"","valueType":"triggerId"}],"x":90,"y":240,"wires":[["221f4e93fadc1fe8"],[]]},{"id":"221f4e93fadc1fe8","type":"api-current-state","z":"96708db28a6dbdc3","name":"Cable Connected?","server":"6fd34f.07042cb","version":3,"outputs":2,"halt_if":"disconnected","halt_if_type":"str","halt_if_compare":"is_not","entity_id":"sensor.driveway_status","state_type":"str","blockInputOverrides":false,"outputProperties":[{"property":"payload","propertyType":"msg","value":"","valueType":"entityState"},{"property":"data","propertyType":"msg","value":"","valueType":"entity"}],"for":"0","forType":"num","forUnits":"minutes","override_topic":false,"state_location":"payload","override_payload":"msg","entity_location":"data","override_data":"msg","x":290,"y":240,"wires":[["aa747b0e13dc8092"],[]]},{"id":"7e0823a129a8ce03","type":"ps-strategy-lowest-price","z":"96708db28a6dbdc3","name":"Lowest Price","fromTime":"18","toTime":"07","hoursOn":"5","doNotSplit":false,"sendCurrentValueWhenRescheduling":true,"outputIfNoSchedule":"false","outputOutsidePeriod":"false","contextStorage":"memory","x":770,"y":140,"wires":[["47448edbd9c76c15"],["ff025927d417b28f"],["031b05ea780117d5"]]},{"id":"6fd34f.07042cb","type":"server","name":"Home Assistant","version":2,"addon":true,"rejectUnauthorizedCerts":true,"ha_boolean":"y|yes|true|on|home|open","connectionDelay":true,"cacheJson":true,"heartbeat":false,"heartbeatInterval":"30"}]`

And this is the message I use to change the config:
`{"payload":{"config":{"fromTime":"06:00:00","toTime":"18:00:00","hoursOn":1,"doNotSplit":false,"sendCurrentValueWhenRescheduling":true,"outputIfNoSchedule":false,"outputOutsidePeriod":false}},"data":{"entity_id":"input_button.update_smart_charging","old_state":{"entity_id":"input_button.update_smart_charging","state":"2022-06-13T06:51:56.288304+00:00","attributes":{"editable":true,"icon":"mdi:ev-station","friendly_name":"Update smart charging"},"last_changed":"2022-06-13T06:51:56.288688+00:00","last_updated":"2022-06-13T06:51:56.288688+00:00","context":{"id":"01G5DXAMZVFJMDYB6S4WSQ8DP0","parent_id":null,"user_id":"6923aa91f7ec459685e9800a5fb3c489"},"original_state":"2022-06-13T06:51:56.288304+00:00"},"new_state":{"entity_id":"input_button.update_smart_charging","state":"2022-06-13T06:54:39.687020+00:00","attributes":{"editable":true,"icon":"mdi:ev-station","friendly_name":"Update smart charging"},"last_changed":"2022-06-13T06:54:39.687180+00:00","last_updated":"2022-06-13T06:54:39.687180+00:00","context":{"id":"01G5DXFMJ5XK69G6GWZ9NB428R","parent_id":null,"user_id":"6923aa91f7ec459685e9800a5fb3c489"},"original_state":"2022-06-13T06:54:39.687020+00:00","timeSinceChangedMs":10}},"_msgid":"382112dd95b3939d"}`

**Comment by ottopaulsen** [comment]
The config message only needs to hold the information you want to change, so if you only want to set the number of hours, this message should be enough:

```json
{
    "payload": {
        "config": {
            "hoursOn": 1
        }
    }
}
```

However, I doubt that is the reason.

Can you look what is stored in the context for the node. Here is an example from me:

<img width="287" alt="image" src="https://user-images.githubusercontent.com/3034647/173417100-7405e8a1-0ed9-464d-8702-7bcab67dc2ff.png">

Verify that the lastPriceData is saved in `file` and that the config is saved in `default`.

Then send a config message, and check the context again. Does it still contain the same data?

**Comment by gulli1986** [comment]
I did some further testing as you mentioned. Sending only the necessary config did not help, however looking at the context data for the node as you proposed I found out it was empty which explains the `No price data`.

As soon as I update the prices trough the `Price Receiver` node, the context data fills in accordingly and the error disappears. My plan was to update the prices as soon as my charging cable was connected to the car. However it looks like it sometimes do not register the connection and hence the prices are not updated in the `Lowest price` node.

As a workaround I now update the prices / schedule as soon as a new dynamic configuration is pushed to the node. It looks like it is working well.

Thanks for you time and the work you do Otto, much appreciated!
